### PR TITLE
Use std::process::Command throughout compile-test

### LINKED ISSUE #43762: Rewrite compiletest to use `Command` liberally
Believe it or not but the `compiletest` project predates the usage of `Command` in the standard library. We currently pass through tons of variables in tons of places, eventually culminating in creating a `Command` at the very end. We should instead thread through a `Command` everywhere to make argument passing much nicer!

That way we also don't need [functions with seven arguments](https://github.com/rust-lang/rust/blob/master/src/tools/compiletest/src/procsrv.rs#L70-L77)!

### PR DESCRIPTION
**Pull request by euclio**
Resubmission of #43798.

Fixes #43762.

r? @alexcrichton

### CODE REVIEW

**Comment by kennytm** [comment]
Almost all GDB tests on `android-arm` failed with "Can't bind address: Address already in use". Possibly related to this PR since the testing code is modified. 

<details>

```
Check compiletest suite=debuginfo mode=debuginfo-gdb (x86_64-unknown-linux-gnu -> arm-linux-androideabi)
[00:58:24] arm-linux-androideabi debug-info test uses tcp 5039 port.please reserve it
[00:58:24] 
[00:58:24] running 108 tests
[00:58:24] test [debuginfo-gdb] debuginfo-gdb/associated-types.rs ... [100%] /data/tmp/work/associated-types.stage2-arm-linux-androideabi
[00:58:24] /checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/associated-types.stage2-arm-linux-androideabi: 1 file pushed. 1.8 MB/s (18292 bytes in 0.010s)
[00:58:24] FAILED
[00:58:24] test [debuginfo-gdb] debuginfo-gdb/basic-types-globals-metadata.rs ... Process /data/tmp/work/associated-types.stage2-arm-linux-androideabi created; pid = 31093
[00:58:24] Listening on port 5039
[00:58:24] [100%] /data/tmp/work/basic-types-globals-metadata.stage2-arm-linux-androideabi
[00:58:24] /checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/basic-types-globals-metadata.stage2-arm-linux-androideabi: 1 file pushed. 1.5 MB/s (12668 bytes in 0.008s)
[00:58:24] FAILED
[00:58:24] test [debuginfo-gdb] debuginfo-gdb/basic-types-globals.rs ... Process /data/tmp/work/basic-types-globals-metadata.stage2-arm-linux-androideabi created; pid = 31098
[00:58:24] Can't bind address: Address already in use.
[00:58:24] Killing process(es): 31098
[00:58:24] [100%] /data/tmp/work/basic-types-globals.stage2-arm-linux-androideabi
[00:58:24] /checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/basic-types-globals.stage2-arm-linux-androideabi: 1 file pushed. 2.0 MB/s (12332 bytes in 0.006s)
[00:58:24] FAILED
<snip>
[00:58:53] test [debuginfo-gdb] debuginfo-gdb/vec.rs ... Process /data/tmp/work/vec-slices.stage2-arm-linux-androideabi created; pid = 31573
[00:58:53] Can't bind address: Address already in use.
[00:58:53] Killing process(es): 31573
[00:58:53] [100%] /data/tmp/work/vec.stage2-arm-linux-androideabi
[00:58:53] /checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/vec.stage2-arm-linux-androideabi: 1 file pushed. 1.6 MB/s (9412 bytes in 0.005s)
[00:58:53] FAILED
[00:58:53] 
[00:58:53] failures:
[00:58:53] 
[00:58:53] ---- [debuginfo-gdb] debuginfo-gdb/associated-types.rs stdout ----
[00:58:53] 	NOTE: compiletest thinks it is using GDB without native rust support
[00:58:53] thread '[debuginfo-gdb] debuginfo-gdb/associated-types.rs' panicked at 'called `Option::unwrap()` on a `None` value', /checkout/src/libcore/option.rs:335:20
[00:58:53] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:58:53] 
[00:58:53] ---- [debuginfo-gdb] debuginfo-gdb/basic-types-globals-metadata.rs stdout ----
[00:58:53] 	NOTE: compiletest thinks it is using GDB without native rust support
[00:58:53] thread '[debuginfo-gdb] debuginfo-gdb/basic-types-globals-metadata.rs' panicked at 'called `Option::unwrap()` on a `None` value', /checkout/src/libcore/option.rs:335:20
[00:58:53] 
<snip>
[00:58:53] ---- [debuginfo-gdb] debuginfo-gdb/vec.rs stdout ----
[00:58:53] 	NOTE: compiletest thinks it is using GDB without native rust support
[00:58:53] thread '[debuginfo-gdb] debuginfo-gdb/vec.rs' panicked at 'called `Option::unwrap()` on a `None` value', /checkout/src/libcore/option.rs:335:20
[00:58:53] 
[00:58:53] 
[00:58:53] failures:
[00:58:53]     [debuginfo-gdb] debuginfo-gdb/associated-types.rs
[00:58:53]     [debuginfo-gdb] debuginfo-gdb/basic-types-globals-metadata.rs
<snip>
[00:58:53]     [debuginfo-gdb] debuginfo-gdb/vec.rs
[00:58:53] 
[00:58:53] test result: FAILED. 0 passed; 98 failed; 10 ignored; 0 measured; 0 filtered out
```

</details>

**Comment by euclio** [comment]
Hm, looking at the GDB and ADB code, I don't see anything glaringly different.

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -1230,41 +1185,39 @@ actual:\n\
...
-                args.prog = self.config.remote_test_client.clone().unwrap()
-                                .into_os_string().into_string().unwrap();
-                self.compose_and_run(args,
-                                     env,
+                let mut test_client = Command::new(
+                    self.config.remote_test_client.as_ref().unwrap());
+                test_client
+                    .args(&["run", &prog])
+                    .args(args)
+                    .envs(env.clone());
```
Comment (Mark-Simulacrum) [inline src/tools/compiletest/src/runtest.rs]: hm, I think passing `&env` here (and in other calls to `envs`) should work, but I'm not certain.

**Comment by Mark-Simulacrum** [comment]
Yeah, not clear to me why this would have failed CI. Could you try running locally and seeing if you can reproduce? `./src/ci/docker/run.sh arm-android`.

**Comment by alexcrichton** [comment]
I think that the stdout for the spawned `adb shell` process isn't captured like it was before, which I think can cause problems like this in this situation.

**Comment by euclio** [comment]
Thanks for the hints. I am on vacation through Tuesday but I'll take a look
when I can.

File: src/tools/compiletest/src/runtest.rs (hunk 1 above)
Comment (euclio) [inline src/tools/compiletest/src/runtest.rs]: This won't work because `envs` requires the `Item` to be a tuple, not a tuple reference.

**Comment by euclio** [comment]
I don't have a computer powerful enough to run the android tests locally for now, but I pushed a commit that reattaches the stdout and stderr for `adb shell`.

**Comment by Mark-Simulacrum** [comment]
You can temporarily add `ALLOW_PR=1` to the android builder so that it runs on Travis and we can get feedback without going through bors.

**Comment by alexcrichton** [comment]
Er the stdout here should be `Stdio::piped` where the stderr is `Stdio::inherit`, we try to capture the stdout just below the call to `spawn`.

**Comment by euclio** [comment]
@alexcrichton Fixed.

**Comment by alexcrichton** [comment]
@bors: r+

Let's see what happens!

**Pull request merged**